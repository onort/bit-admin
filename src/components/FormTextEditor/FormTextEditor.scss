@import "../../styles/main.scss";

// Default HTML heading, ul and blockquite stlyes to revert minireset
h1 {
  display: block;
  font-size: 2em;
  margin-top: 0.67em;
  margin-bottom: 0.67em;
  margin-left: 0;
  margin-right: 0;
  font-weight: bold;
}

h2 {
  display: block;
  font-size: 1.5em;
  margin-top: 0.83em;
  margin-bottom: 0.83em;
  margin-left: 0;
  margin-right: 0;
  font-weight: bold;
}

h3 {
  display: block;
  font-size: 1.17em;
  margin-top: 1em;
  margin-bottom: 1em;
  margin-left: 0;
  margin-right: 0;
  font-weight: bold;
}

h4 {
  display: block;
  font-size: 1em;
  margin-top: 1.33em;
  margin-bottom: 1.33em;
  margin-left: 0;
  margin-right: 0;
  font-weight: bold;
}

h5 {
  display: block;
  font-size: 0.83em;
  margin-top: 1.67em;
  margin-bottom: 1.67em;
  margin-left: 0;
  margin-right: 0;
  font-weight: bold;
}

h6 {
  display: block;
  font-size: 0.67em;
  margin-top: 2.33em;
  margin-bottom: 2.33em;
  margin-left: 0;
  margin-right: 0;
  font-weight: bold;
}

ul {
  list-style-type: disc;
}

blockquote {
  display: block;
  margin-top: 1em;
  margin-bottom: 1em;
  margin-left: 40px;
  margin-right: 40px;
}

// Editor Styles

.label {
  display: block;
  font-size: $font-size-md;
  font-weight: 600;
  letter-spacing: 0.66px;
  margin-bottom: $spacer-xxsm;
}

.container {
  background-color: $white;
  border: 1px solid $light-border;
  border-radius: px-to-rem(4px);
  padding: $spacer-sm $spacer-xsm;
  margin: $spacer-sm 0;
}

.labeled {
  margin-top: 0;
}

.editor {
  border-top: 1px solid $light-border;
  margin-top: $spacer-sm;
  padding-top: $spacer-sm;
  min-height: px-to-rem(140px);
}

// StyleButton Component Styles
.controlRow {
  display: flex;
  margin-bottom: $spacer-xsm;
}

.styleButton {
  display: inline-block;
  margin-right: $spacer-xsm;
  padding: px-to-rem(3px) px-to-rem(6px);
  background-color: $white;
  font-size: $font-size-sm;
  color: $light-grey;
  border: 1px solid lighten($light-grey, 40);
  border-radius: px-to-rem(2px);
  cursor: pointer;
  transition: all 0.15s;

  &:hover {
    background-color: lighten($orange, 40);
    border-color: $orange;
    color: $orange;
  }
}

.activeButton {
  color: $orange;
  border-color: $orange;

  &:hover {
    background-color: $white;
    color: $orange;
    border-color: $orange;
  }
}
