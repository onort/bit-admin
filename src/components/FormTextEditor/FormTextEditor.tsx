import React from "react"
import { FieldProps } from "formik"
import {
  Editor,
  EditorState,
  RichUtils,
  DraftBlockType,
  convertToRaw,
  DraftInlineStyleType
} from "draft-js"
import "draft-js/dist/Draft.css"
import cx from "classnames"

import styles from "./FormTextEditor.scss"
import { BlockStyles, InlineStyles } from "./"

interface Props extends FieldProps {
  className?: string
  label?: string
}

const FormTextEditor: React.SFC<Props> = props => {
  const { field, form } = props

  const handleChange = (editorState: EditorState) => {
    form.setFieldValue(field.name, editorState)
    // console.log("convertToRaw", convertToRaw(editorState.getCurrentContent()))
  }

  const handleBlockTypeToggle = (blockType: DraftBlockType) =>
    handleChange(RichUtils.toggleBlockType(field.value, blockType))

  const handleInlineStyleTypeToggle = (inlineStyle: DraftInlineStyleType) =>
    handleChange(RichUtils.toggleInlineStyle(field.value, inlineStyle))

  const className = cx(styles.container, { [styles.labeled]: props.label })

  return (
    <>
      {props.label && (
        <label className={styles.label} htmlFor={name}>
          {props.label}
        </label>
      )}
      <div className={className}>
        <BlockStyles
          editorState={field.value}
          onToggle={handleBlockTypeToggle}
        />
        <InlineStyles
          editorState={field.value}
          onToggle={handleInlineStyleTypeToggle}
        />
        <div className={styles.editor}>
          <Editor
            editorState={field.value}
            onChange={handleChange}
            placeholder="Add content..."
          />
        </div>
      </div>
    </>
  )
}

export default FormTextEditor
