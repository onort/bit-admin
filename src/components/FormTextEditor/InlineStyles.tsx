import React from "react"
import { EditorState, DraftInlineStyleType } from "draft-js"

import styles from "./FormTextEditor.scss"
import { StyleButton } from "./"

interface Props {
  editorState: EditorState
  onToggle: (inlineStyle: DraftInlineStyleType) => void
}

const INLINE_STYLES = [
  { label: "Bold", style: "BOLD" },
  { label: "Italic", style: "ITALIC" },
  { label: "Underline", style: "UNDERLINE" },
  { label: "Monospace", style: "CODE" }
]

const BlockStyleControls: React.SFC<Props> = props => {
  const { editorState, onToggle } = props
  const currentStyle = editorState.getCurrentInlineStyle()

  return (
    <div className={styles.controlRow}>
      {INLINE_STYLES.map(type => (
        <StyleButton
          key={type.label}
          active={currentStyle.has(type.style)}
          label={type.label}
          onToggle={onToggle}
          style={type.style}
        />
      ))}
    </div>
  )
}

export default BlockStyleControls
