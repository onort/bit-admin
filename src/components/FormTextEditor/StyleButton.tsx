import React from "react"
import { DraftBlockType, DraftInlineStyleType } from "draft-js"
import cx from "classnames"

import styles from "./FormTextEditor.scss"

interface Props {
  active: boolean
  label: string
  onToggle: (e: any) => void
  style: string
}

const StyleButton: React.SFC<Props> = props => {
  const onToggle = (e: React.MouseEvent) => {
    e.preventDefault()
    props.onToggle(props.style)
  }
  const className = cx(styles.styleButton, {
    [styles.activeButton]: props.active
  })
  return (
    <span className={className} onMouseDown={onToggle}>
      {props.label}
    </span>
  )
}

export default StyleButton
