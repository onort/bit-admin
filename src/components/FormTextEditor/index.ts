import FormTextEditor from "./FormTextEditor"
import BlockStyles from "./BlockStyles"
import InlineStyles from "./InlineStyles"
import StyleButton from "./StyleButton"

export default FormTextEditor

export { BlockStyles, InlineStyles, StyleButton }
