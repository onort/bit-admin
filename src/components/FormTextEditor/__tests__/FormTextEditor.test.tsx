import React from "react"
import { mount, shallow } from "enzyme"
import { Formik, Form, Field } from "formik"
import { EditorState, Editor } from "draft-js"

import FormTextEditor, { BlockStyles, InlineStyles } from "../"

describe("<FormTextEditor />", () => {
  // random generated keys breaks snapshot tests
  it.skip("should match snapshot", () => {
    jest.mock("draft-js/lib/generateRandomKey.js", () => () => "123")
    const testValues = { testEditorState: EditorState.createEmpty() }
    const fakeSubmit = jest.fn()
    const wrapper = shallow(
      <Formik initialValues={testValues} onSubmit={fakeSubmit}>
        {props => {
          return (
            <form onSubmit={props.handleSubmit}>
              <Field
                name="testEditorState"
                label="Test Text Editor"
                component={FormTextEditor}
              />
            </form>
          )
        }}
      </Formik>
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should render editor", () => {
    const testValues = { testEditorState: EditorState.createEmpty() }
    const fakeSubmit = jest.fn()
    const wrapper = mount(
      <Formik initialValues={testValues} onSubmit={fakeSubmit}>
        {props => {
          return (
            <form onSubmit={props.handleSubmit}>
              <Field
                name="testEditorState"
                label="Test Text Editor"
                component={FormTextEditor}
              />
            </form>
          )
        }}
      </Formik>
    )
    expect(wrapper.find(Editor).exists()).toBe(true)
    wrapper.unmount()
  })

  it("should have blockstyle controls", () => {
    const testValues = { testEditorState: EditorState.createEmpty() }
    const fakeSubmit = jest.fn()
    const wrapper = mount(
      <Formik initialValues={testValues} onSubmit={fakeSubmit}>
        {props => {
          return (
            <form onSubmit={props.handleSubmit}>
              <Field
                name="testEditorState"
                label="Test Text Editor"
                component={FormTextEditor}
              />
            </form>
          )
        }}
      </Formik>
    )
    expect(wrapper.find(BlockStyles).exists()).toBe(true)
    wrapper.unmount()
  })

  it("should have inlinestyle controls", () => {
    const testValues = { testEditorState: EditorState.createEmpty() }
    const fakeSubmit = jest.fn()
    const wrapper = mount(
      <Formik initialValues={testValues} onSubmit={fakeSubmit}>
        {props => {
          return (
            <form onSubmit={props.handleSubmit}>
              <Field
                name="testEditorState"
                label="Test Text Editor"
                component={FormTextEditor}
              />
            </form>
          )
        }}
      </Formik>
    )
    expect(wrapper.find(InlineStyles).exists()).toBe(true)
    wrapper.unmount()
  })
})
