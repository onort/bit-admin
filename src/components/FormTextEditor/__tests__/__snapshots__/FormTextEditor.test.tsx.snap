// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormTextEditor /> should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "dirty": false,
      "errors": Object {},
      "handleBlur": [Function],
      "handleChange": [Function],
      "handleReset": [Function],
      "handleSubmit": [Function],
      "initialValues": Object {
        "testEditorState": EditorState {
          "_immutable": Immutable.Record {
            "allowUndo": true,
            "currentContent": Immutable.Record {
              "entityMap": Object {
                "__add": [Function],
                "__create": [Function],
                "__get": [Function],
                "__getLastCreatedEntityKey": [Function],
                "__mergeData": [Function],
                "__replaceData": [Function],
                "add": [Function],
                "create": [Function],
                "get": [Function],
                "getLastCreatedEntityKey": [Function],
                "mergeData": [Function],
                "replaceData": [Function],
              },
              "blockMap": Immutable.OrderedMap {
                "1kbu8": Immutable.Record {
                  "key": "1kbu8",
                  "type": "unstyled",
                  "text": "",
                  "characterList": Immutable.List [],
                  "depth": 0,
                  "data": Immutable.Map {},
                },
              },
              "selectionBefore": Immutable.Record {
                "anchorKey": "1kbu8",
                "anchorOffset": 0,
                "focusKey": "1kbu8",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
              "selectionAfter": Immutable.Record {
                "anchorKey": "1kbu8",
                "anchorOffset": 0,
                "focusKey": "1kbu8",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
            },
            "decorator": null,
            "directionMap": Immutable.OrderedMap {
              "1kbu8": "LTR",
            },
            "forceSelection": false,
            "inCompositionMode": false,
            "inlineStyleOverride": null,
            "lastChangeType": null,
            "nativelyRenderedContent": null,
            "redoStack": Immutable.Stack [],
            "selection": Immutable.Record {
              "anchorKey": "1kbu8",
              "anchorOffset": 0,
              "focusKey": "1kbu8",
              "focusOffset": 0,
              "isBackward": false,
              "hasFocus": false,
            },
            "treeMap": Immutable.OrderedMap {
              "1kbu8": Immutable.List [
                Immutable.Record {
                  "start": 0,
                  "end": 0,
                  "decoratorKey": null,
                  "leaves": Immutable.List [
                    Immutable.Record {
                      "start": 0,
                      "end": 0,
                    },
                  ],
                },
              ],
            },
            "undoStack": Immutable.Stack [],
          },
        },
      },
      "isSubmitting": false,
      "isValid": false,
      "isValidating": false,
      "registerField": [Function],
      "resetForm": [Function],
      "setError": [Function],
      "setErrors": [Function],
      "setFieldError": [Function],
      "setFieldTouched": [Function],
      "setFieldValue": [Function],
      "setFormikState": [Function],
      "setStatus": [Function],
      "setSubmitting": [Function],
      "setTouched": [Function],
      "setValues": [Function],
      "submitCount": 0,
      "submitForm": [Function],
      "touched": Object {},
      "unregisterField": [Function],
      "validate": undefined,
      "validateField": [Function],
      "validateForm": [Function],
      "validateOnBlur": true,
      "validateOnChange": true,
      "validationSchema": undefined,
      "values": Object {
        "testEditorState": EditorState {
          "_immutable": Immutable.Record {
            "allowUndo": true,
            "currentContent": Immutable.Record {
              "entityMap": Object {
                "__add": [Function],
                "__create": [Function],
                "__get": [Function],
                "__getLastCreatedEntityKey": [Function],
                "__mergeData": [Function],
                "__replaceData": [Function],
                "add": [Function],
                "create": [Function],
                "get": [Function],
                "getLastCreatedEntityKey": [Function],
                "mergeData": [Function],
                "replaceData": [Function],
              },
              "blockMap": Immutable.OrderedMap {
                "1kbu8": Immutable.Record {
                  "key": "1kbu8",
                  "type": "unstyled",
                  "text": "",
                  "characterList": Immutable.List [],
                  "depth": 0,
                  "data": Immutable.Map {},
                },
              },
              "selectionBefore": Immutable.Record {
                "anchorKey": "1kbu8",
                "anchorOffset": 0,
                "focusKey": "1kbu8",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
              "selectionAfter": Immutable.Record {
                "anchorKey": "1kbu8",
                "anchorOffset": 0,
                "focusKey": "1kbu8",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
            },
            "decorator": null,
            "directionMap": Immutable.OrderedMap {
              "1kbu8": "LTR",
            },
            "forceSelection": false,
            "inCompositionMode": false,
            "inlineStyleOverride": null,
            "lastChangeType": null,
            "nativelyRenderedContent": null,
            "redoStack": Immutable.Stack [],
            "selection": Immutable.Record {
              "anchorKey": "1kbu8",
              "anchorOffset": 0,
              "focusKey": "1kbu8",
              "focusOffset": 0,
              "isBackward": false,
              "hasFocus": false,
            },
            "treeMap": Immutable.OrderedMap {
              "1kbu8": Immutable.List [
                Immutable.Record {
                  "start": 0,
                  "end": 0,
                  "decoratorKey": null,
                  "leaves": Immutable.List [
                    Immutable.Record {
                      "start": 0,
                      "end": 0,
                    },
                  ],
                },
              ],
            },
            "undoStack": Immutable.Stack [],
          },
        },
      },
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <C
      component={[Function]}
      label="Test Text Editor"
      name="testEditorState"
    />
  </form>
</ContextProvider>
`;

exports[`<FormTextEditor /> should match snapshot 2`] = `"<form><label class=\\"label\\" for=\\"nodejs\\">Test Text Editor</label><div class=\\"container labeled\\"><div class=\\"controlRow\\"><span class=\\"styleButton\\">H1</span><span class=\\"styleButton\\">H2</span><span class=\\"styleButton\\">H3</span><span class=\\"styleButton\\">H4</span><span class=\\"styleButton\\">H5</span><span class=\\"styleButton\\">H6</span><span class=\\"styleButton\\">Blockquote</span><span class=\\"styleButton\\">UL</span><span class=\\"styleButton\\">OL</span><span class=\\"styleButton\\">Code Block</span></div><div class=\\"controlRow\\"><span class=\\"styleButton\\">Bold</span><span class=\\"styleButton\\">Italic</span><span class=\\"styleButton\\">Underline</span><span class=\\"styleButton\\">Monospace</span></div><div class=\\"editor\\"><div class=\\"DraftEditor-root\\"><div class=\\"public-DraftEditorPlaceholder-root\\"><div class=\\"public-DraftEditorPlaceholder-inner\\" id=\\"placeholder-2fesl\\" style=\\"white-space:pre-wrap\\">Add content...</div></div><div class=\\"DraftEditor-editorContainer\\"><div aria-describedby=\\"placeholder-2fesl\\" class=\\"notranslate public-DraftEditor-content\\" contenteditable=\\"true\\" role=\\"textbox\\" spellcheck=\\"false\\" style=\\"outline:none;user-select:text;-webkit-user-select:text;white-space:pre-wrap;word-wrap:break-word\\"><div data-contents=\\"true\\"><div class=\\"\\" data-block=\\"true\\" data-editor=\\"2fesl\\" data-offset-key=\\"1kbu8-0-0\\"><div data-offset-key=\\"1kbu8-0-0\\" class=\\"public-DraftStyleDefault-block public-DraftStyleDefault-ltr\\"><span data-offset-key=\\"1kbu8-0-0\\"><br data-text=\\"true\\"/></span></div></div></div></div></div></div></div></div></form>"`;
